### ISSUE TITLE
Generic Oracle Requirements 

### ISSUE THREAD

**Issue by echevrier** [integritee-network/teeracle#15]
The teeracle is a service that provides trustworthy oracles to the integritee-parachain. It retrieves website or API data and serves source-authenticated data to the parachain.

# Stakeholder:
* Dapp (smart contract or extrinsic sender) who need a trusted oracle but may not want to operate their own oracle infrastructure
* Teeracle Operator who wants to earn money

# User Story
Someone wants to run a decentralized application that needs a trusted oracle to gain information, but they don't want to operate their own oracle infrastructure. They will choose an oracle operator and will pay him fees to access the data.

Example:

- A weather forecasts Dapp
- A flight delay insurance Dapp. See use case in Epic https://github.com/integritee-network/worker/issues/1228

# Requirements

- The Teeracle offers the service of executing custom oracle jobs:
   - Custom data request: source of the data, content to be retrieved and returned data type
   - Custom schedule: queries frequency and duration.
- The Teeracle provides source-authenticated data and transmit the data faithfully.
- The Teeracle can use API credentials confidentially per request. (allow the use of keys for paid APIs)
- The requester can order a custom oracle job to be executed on a specific number of Teeracles. 
- The oracle job can be defined at runtime.
- A requested oracle job can be cancelled. payment is due pro-rata in this case.
- The returned data must be fed into the Teeracle Pallet. Later it will be perhaps also fed into other blockchains.
- The returned data should be timestamped.
- The Teeracle operator will charge a fee in TEER tokens for the use of Teeracle. This fee will only be paid if the requested data is successfully published onchain. 
- The Teeracle usage fees are defined chain-wide on a per-result-publishing and per result-byte basis. it cannot be less than the fee which should be paid for sending the extrinsic with the query result at the requested frequency.
- The Teeracle will pay out earnings from its signing key to its operator account. operator can specify the beneficiary account, a threshold which triggers payout and a withdrawal amount. 
- The implementation must be so generic, that other use case can be easily implemented:
   - Trusted exchange rate oracle for a trading pair
   - Stable Fees for the Integritee Parachain. See: https://github.com/integritee-network/pallets/issues/34
   - A flight insurance contract. See https://github.com/integritee-network/worker/issues/1228
   - ...

# Non-requirements
* It seems that there is no replay protection in TLS.  solving this is out of scope

# Specification

## jobs
A job shall be sent to the teeracle pallet either 
* through an extrinsic 
* originating from a smart contract on the integritee-parachain
* XCM from another parachain

A job shall contain the following fields
* request url
* request parameters
* (optional) api key to use, encrypted using the teeracle shielding key
* result parsing instruction (assume json result formatting for now)
* return-type definition (must be bounded in size)
* schedule
   * how many requests total
   * request frequency (number of blocks between requests)
   * how many teeracles shall answer each request in parallel

## teeracle pallet

*naming is tentative*

* provides a dispatchables:
   * register_job (total fee for the entire schedule will be reserved)
   * cancel_job
   * request refund (if teeracles don't answer within timeout, the requester can trigger release of remaining reservation )
   * register_result (ensure is whitelisted teeracle and registered in teerex and assigned to the job) trigger payout of fees to teeracle (signing key) accounts
* defines a timeout for requests to be answered by the teeracles
* randomly assigns teeracle instances to jobs
